client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,320 @@
+        this.timeoutFuture = executor.scheduleWithFixedDelay(new TimeoutBatch(encoder),
+                                                             MAX_BATCH_TIME_MILLIS,
+                                                             MAX_BATCH_TIME_MILLIS,
+                                                             TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    static class ConnectionReader implements Runnable {
+        static final ThreadFactory THREAD_FACTORY = ExecutorServiceHelpers.getThreadFactory("ClientSocketReaders", (Thread.NORM_PRIORITY + Thread.MAX_PRIORITY) / 2);
+        
+        private final String name;
+        private final InputStream in;
+        private final ReplyProcessor callback;
+        private final Thread thread;
+        private final AppendBatchSizeTracker batchSizeTracker;
+        private final AtomicBoolean stop = new AtomicBoolean(false);
+
+        public ConnectionReader(String name, InputStream in, ReplyProcessor callback,
+                                AppendBatchSizeTracker batchSizeTracker) {
+            this.name = name;
+            this.in = in;
+            this.callback = callback;
+            this.thread = THREAD_FACTORY.newThread(this);
+            this.batchSizeTracker = batchSizeTracker;
+        }
+        
+        public void start() {
+            thread.start();
+        }
+        
+        public void run() {
+            IoBuffer buffer = new IoBuffer();
+            while (!stop.get()) {
+                try {
+                    WireCommand command = readCommand(in, buffer);
+                    if (command instanceof WireCommands.DataAppended) {
+                        WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
+                        batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                    }
+                    try {
+                        callback.process((Reply) command);
+                    } catch (Exception e) {
+                        callback.processingFailure(e);
+                    }
+                } catch (SocketException e) {
+                    if (e.getMessage().equals("Socket closed")) {
+                        log.info("Closing TcpConnection.Reader because socket is closed.");
+                    } else {
+                        log.warn("Error in reading from socket.", e);
+                    }
+                    stop();
+                } catch (EOFException e) {
+                    log.info("Closing TcpConnection.Reader because end of input readched.");
+                    stop();
+                } catch (Exception e) {
+                    log.warn("Error processing data from from server " + name, e);
+                    stop();
+                }
+            }
+        }
+
+        @VisibleForTesting
+        static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOException {
+            ByteBuf header = buffer.getBuffOfSize(in, 8);
```

Comment:
here we can use ByteBuf and recycle it.
Using the ByteBufAllocator is the simpler way.
it will make code simpler, without the need of the IoBuffer class

Reply:
My benchmarks showed that doing that was slower.

Reply:
Interesting.
Maybe the problem for netty is that we passing thru a generic InputStream and probably we can't leverage zero copy features.
If we had access to the SocketChannel maybe we could skip the copies to the heap.

Btw it is something we can do in some follow up work.

I am fine with this approach currently